lib/rouge/lexers/php.rb

```diff
@@ -65,14 +68,17 @@ def builtins
       def self.keywords
         @keywords ||= Set.new %w(
           and E_PARSE old_function E_ERROR or as E_WARNING parent eval
-          PHP_OS break exit case extends PHP_VERSION cfunction FALSE
+          PHP_OS break exit case extends PHP_VERSION cfunction false
           print for require continue foreach require_once declare return
-          default static do switch die stdClass echo else TRUE elseif
+          default static do switch die stdclass echo else true elseif
           var empty if xor enddeclare include virtual endfor include_once
-          while endforeach global __FILE__ endif list __LINE__ endswitch
-          new __sleep endwhile not array __wakeup E_ALL NULL final
+          while endforeach global __file__ endif list __line__ endswitch
```

Comment:
Similar question with respect to these directives as with `stdclass`.

Reply:
Pull out these E_\* and PHP_\* predefined constants apart as a new instance variable?

```
def self.constants
  @constants ||= Set.new %w(
    E_ERROR
    E_RECOVERABLE_ERROR
    E_WARNING
    E_PARSE
    E_NOTICE
    E_STRICT
    E_DEPRECATED
    E_CORE_ERROR
    E_CORE_WARNING
    E_COMPILE_ERROR
    E_COMPILE_WARNING
    E_USER_ERROR
    E_USER_WARNING
    E_USER_NOTICE
    E_USER_DEPRECATED
    E_ALL
    # to be completed with PHP_*
  )
end
```

Or directly by a new rule? (`rule %r/E_(?:(?:(?:CORE|COMPILE|USER)_)?(?:ERROR|WARNING)|(?:(?:USER_)?(?:NOTICE|DEPRECATED))|(?:RECOVERABLE_ERROR|PARSE|STRICT|ALL))/, Name::Constant`)

Reply:
Is that how short the rule could have to be? It doesn't look that bad.

Reply:
This is just for the E\_\* constants, to illustrate, I have around 50 PHP\_\* constants to handle/add.

Reply:
It's a non-goal of Rouge to check for syntax correctness. Could we simplify things by having a simple rule that treats identifiers that begin with `E_` as constants? Should fully capitalised identifiers be treated as constants regardless? Basically where I'm coming from is that I would like to reduce the number of special objects we need to create.

Reply:
Same story here. My apologies for not making sure I remembered properly what we had been talking about. I'm in favour of removing the `E_` constants referred to here. What I was trying to get at in the previous comment was, if you were to do that, could you have a simple rule that merely matched identifiers of the form `/E_[A-Z]+/` and applied a token to those matches?

Reply:
I like your suggestion, I think it's a "fair" compromise and simpler.